### PR TITLE
filter out init containers in GetContainers()

### PR DESCRIPTION
**Pull request by yangcao77**
Signed-off-by: Stephanie <yangcao@redhat.com>

### What does this PR do?:

This PR filters out initContainers in `GetContainers()`

### Which issue(s) this PR fixes:

Fixes https://github.com/devfile/api/issues/532

### PR acceptance criteria:
Testing and documentation do not need to be complete in order for this PR to be approved. We just need to ensure tracking issues are opened.

> - Open new test/doc issues under the [devfile/api](https://github.com/devfile/api/issues) repo
> - Check each criteria if:
>  - There is a separate tracking issue. Add the issue link under the criteria
>  **or**
>  - test/doc updates are made as part of this PR
> -  If unchecked, explain why it's not needed

- [x] Unit/Functional tests

  

- [ ] [QE Integration test](https://github.com/devfile/integration-tests) 

  

- [ ] Documentation 

   

- [ ] Client Impact

  

### How to test changes / Special notes to the reviewer:
call `GetContainers` against a devfile with init container, the output is expected to return all containers other than init containers

### CODE REVIEW

File: pkg/devfile/generator/generators.go [hunk 1]
```diff
@@ -55,54 +55,50 @@ func GetObjectMeta(name, namespace string, labels, annotations map[string]string
...
-			err = addSyncFolder(container, syncRootFolder, projects)
-			if err != nil {
-				return nil, err
+		for _, event := range preStartEvents {
+			eventSubCommands := common.GetCommandsFromEvent(commandsMap, event)
+			eventCommands = append(eventCommands, eventSubCommands...)
+		}
+
+		for _, commandName := range eventCommands {
+			if command, ok := commandsMap[commandName]; ok {
```
Comment (maysunfaisal) [inline pkg/devfile/generator/generators.go:82]: this check is really not needed, its guaranteed to be in commands Map via `GetCommandsFromEvent()`

Comment (yangcao77) [reply pkg/devfile/generator/generators.go]: you are right. updated

File: pkg/devfile/generator/utils_test.go [hunk 2]
```diff
@@ -774,7 +774,7 @@ func TestGetServiceSpec(t *testing.T) {
 				mockGetComponents.Return(tt.filteredComponents, nil).AnyTimes()
 			}
 			mockDevfileData.EXPECT().GetProjects(common.DevfileOptions{}).Return(nil, nil).AnyTimes()
-
+			mockDevfileData.EXPECT().GetEvents().Return(v1.Events{}).AnyTimes()
```
Comment (maysunfaisal) [inline pkg/devfile/generator/utils_test.go:775]: going thru this, I want to confirm with u if `getServiceSpec` should call `GetContainers`. We only want container ports from the main containers right?

File: pkg/devfile/generator/generators.go [hunk 3]
```diff
@@ -55,54 +55,49 @@ func GetObjectMeta(name, namespace string, labels, annotations map[string]string
...
-			Image:        comp.Container.Image,
-			IsPrivileged: false,
-			Command:      comp.Container.Command,
-			Args:         comp.Container.Args,
-			EnvVars:      envVars,
-			ResourceReqs: resourceReqs,
-			Ports:        ports,
+
+	// filter out init containers
+	preStartEvents := devfileObj.Data.GetEvents().PreStart
```
Comment (maysunfaisal) [inline pkg/devfile/generator/generators.go:66]: how about PostStop?

PostStop are also Apply commands that can reference a component container and they are not the main containers i.e.; PostStop is executed after the main components are brought down 🤔

Comment (yangcao77) [reply pkg/devfile/generator/generators.go:66]: postStop event is not for  init containers tho.

File: pkg/devfile/generator/utils_test.go (hunk 2 above)
Comment (yangcao77) [reply pkg/devfile/generator/utils_test.go:775]: Yes.  it is calling`GetContainers`, which only returns main containers.

File: pkg/devfile/generator/generators.go (hunk 3 above)
Comment (maysunfaisal) [reply pkg/devfile/generator/generators.go:66]: but postStop are probably not part of the main containers too 🤔

Comment (yangcao77) [reply pkg/devfile/generator/generators.go:66]: you are right. updated `GetContainers()`, and also updated the test to filter out containers for preStart and postStop events.

File: pkg/devfile/generator/generators_test.go [hunk 4]
```diff
@@ -280,12 +280,13 @@ func TestGetContainers(t *testing.T) {
 			// to set up the prestartevent and apply command for init container
 			mockGetCommands := mockDevfileData.EXPECT().GetCommands(common.DevfileOptions{})
 			mockGetCommands.Return(applyCommands, nil).AnyTimes()
-			preStartEvents := v1.Events{
+			events := v1.Events{
 				DevWorkspaceEvents: v1.DevWorkspaceEvents{
 					PreStart: tt.eventCommands,
+					PostStop: tt.eventCommands,
```
Comment (maysunfaisal) [inline pkg/devfile/generator/generators_test.go:286]: last small nit: can you make a separate command for PostStop? Both the events are referencing the same command, but want to make sure we are testing if its removed properly

Comment (yangcao77) [reply pkg/devfile/generator/generators_test.go:286]: updated

**Review by maysunfaisal — APPROVED** [review]

**Pull request merged**